internal/sorter/sorter.go

```diff
@@ -296,33 +297,44 @@ func (s routeSorter) Less(i, j int) bool {
 	switch a := s[i].PathMatchCondition.(type) {
 	case *dag.PrefixMatchCondition:
 		if b, ok := s[j].PathMatchCondition.(*dag.PrefixMatchCondition); ok {
-			cmp := strings.Compare(a.Prefix, b.Prefix)
-			switch cmp {
-			case 1:
+			switch {
+			case len(a.Prefix) > len(b.Prefix):
 				// Sort longest prefix first.
 				return true
-			case -1:
+			case len(a.Prefix) < len(b.Prefix):
 				return false
 			default:
-				if a.PrefixMatchType == b.PrefixMatchType {
-					return compareRoutesByMethodHeaderQueryParams(s[i], s[j])
+				cmp := strings.Compare(a.Prefix, b.Prefix)
+				switch cmp {
+				case 1:
+					return true
+				case -1:
+					return false
+				default:
+					if a.PrefixMatchType == b.PrefixMatchType {
+						v, _ := compareRoutesByMethodHeaderQueryParams(s[i], s[j])
+						return v
+					}
+					// Segment prefixes sort first as they are more specific.
+					return a.PrefixMatchType == dag.PrefixMatchSegment
 				}
-				// Segment prefixes sort first as they are more specific.
-				return a.PrefixMatchType == dag.PrefixMatchSegment
 			}
 		}
 	case *dag.RegexMatchCondition:
 		switch b := s[j].PathMatchCondition.(type) {
 		case *dag.RegexMatchCondition:
-			cmp := strings.Compare(a.Regex, b.Regex)
-			switch cmp {
-			case 1:
+			switch {
+			case len(a.Regex) > len(b.Regex):
 				// Sort longest regex first.
 				return true
-			case -1:
+			case len(a.Regex) < len(b.Regex):
 				return false
 			default:
-				return compareRoutesByMethodHeaderQueryParams(s[i], s[j])
+				v, err := compareRoutesByMethodHeaderQueryParams(s[i], s[j])
```

Comment:
found another test case that does not consistently sort with this implementation but did with the previous (/pathregex2 always sorted before all the /pathregex1 matches):
```diff
diff --git a/internal/sorter/sorter_test.go b/internal/sorter/sorter_test.go
index aac862a4..3958bca4 100644
--- a/internal/sorter/sorter_test.go
+++ b/internal/sorter/sorter_test.go
@@ -392,25 +392,31 @@ func TestSortRoutesLongestHeaders(t *testing.T) {
                        PathMatchCondition: matchExact("/pathexact"),
                },
                {
-                       PathMatchCondition: matchRegex("/pathregex"),
+                       PathMatchCondition: matchRegex("/pathregex1"),
                        HeaderMatchConditions: []dag.HeaderMatchCondition{
                                exactHeader("header-name", "header-value"),
                        },
                },
                {
-                       PathMatchCondition: matchRegex("/pathregex"),
+                       PathMatchCondition: matchRegex("/pathregex1"),
                        HeaderMatchConditions: []dag.HeaderMatchCondition{
                                presentHeader("header-name"),
                        },
                },
                {
-                       PathMatchCondition: matchRegex("/pathregex"),
+                       PathMatchCondition: matchRegex("/pathregex1"),
                        HeaderMatchConditions: []dag.HeaderMatchCondition{
                                exactHeader("long-header-name", "long-header-value"),
                        },
                },
                {
-                       PathMatchCondition: matchRegex("/pathregex"),
+                       PathMatchCondition: matchRegex("/pathregex2"),
+                       HeaderMatchConditions: []dag.HeaderMatchCondition{
+                               presentHeader("header-name"),
+                       },
+               },
+               {
+                       PathMatchCondition: matchRegex("/pathregex1"),
                },
                {
                        PathMatchCondition: matchPrefixSegment("/path"),
@@ -718,6 +724,6 @@ func shuffleAndCheckSort[T any](t *testing.T, want []T) {
                have := shuffleSlice(want)

                sort.Stable(For(have))
-               assert.Equal(t, want, have)
+               assert.Equalf(t, want, have, "failed trial #%d", i)
        }
 }
```

Reply:
I would expect this case `compareRoutesByMethodHeaderQueryParams` to return an error

Reply:
and to fallback to lexi sorting

Reply:
you can imagine another example of regexes of the same length where you don't want the matches to be "interleaved" like could happen with what we have in this PR (e.g. say path matches for `/foo.*` and `/.*foo` with/without header matches)

I was able to get the test to consistently sort (and IMO sort in an order that makes a little more sense) by doing the lexical comparison before comparing header/etc. matches
```diff
diff --git a/internal/sorter/sorter.go b/internal/sorter/sorter.go
index 3a6f4de1..d2d65b69 100644
--- a/internal/sorter/sorter.go
+++ b/internal/sorter/sorter.go
@@ -330,11 +330,16 @@ func (s routeSorter) Less(i, j int) bool {
                        case len(a.Regex) < len(b.Regex):
                                return false
                        default:
-                               v, err := compareRoutesByMethodHeaderQueryParams(s[i], s[j])
-                               if err != nil {
-                                       return strings.Compare(a.Regex, b.Regex) == -1
+                               cmp := strings.Compare(a.Regex, b.Regex)
+                               switch cmp {
+                               case 1:
+                                       return true
+                               case -1:
+                                       return false
+                               default:
+                                       v, _ := compareRoutesByMethodHeaderQueryParams(s[i], s[j])
+                                       return v
                                }
-                               return v
                        }
                case *dag.PrefixMatchCondition:
                        return true
```
(I think you can get rid of the added error return too since we should never get to a case where the route match conditions are fully exactly equal)

Reply:
> you can imagine another example of regexes of the same length where you don't want the matches to be "interleaved" like could happen with what we have in this PR (e.g. say path matches for /foo.* and /.*foo with/without header matches)

This is how I had initially but I thought that we wanted `compareRoutesByMethodHeaderQueryParams` for more specificity first

Reply:
But I agree this change helps quite a bit with stabilizing the sorting algo

Reply:
fixed in b7f8c9eff14519b21bdf30c927b89637a288649c